Fix creation of duplicate tokens

wjrosa opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes #2691

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

This is the first part of a series of fixes to address #2691. Here's a fix for the "add payment method" feature on the "my account" page and checkout (new checkout experience).

When adding a new payment method, we check if any existing token contains the same details. If so, we will then update the existing token with the new payment method ID (from Stripe). If not, we will proceed with creating a new token.

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

- Checkout and build this branch on your test environment (`fix/duplicate-credit-card-tokens`)
- Disable the legacy checkout experience in settings (`wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`)

### Add payment method page
- As a shopper, access your "Payment methods" page (`my-account/payment-methods/`)
- Add a new payment method (ie. the test card ending with 4242)
- Confirm the card is successfully added
- Check the database for the card details and notice the payment method ID, starting with `pm_` (`wp_woocommerce_payment_tokens` table,  `token` column)
- Add the same card again
- Confirm that no new card appears
- Check the database and confirm that just the `token` column was updated with a new payment method ID

### Block checkout
- As a merchant, turn on "Enable payments via saved cards" in `wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe&panel=settings`
- As a shopper, add any product to your cart
- Go to the checkout page
- Select/input a new payment method (ie. the test card ending with 4242)
- Check the option "Save payment information to my account for future purchases."
- Complete the purchase
- Confirm the card is successfully added to your account (check your "My Account" page)
- Check the database for the card details and notice the payment method ID, starting with `pm_` (`wp_woocommerce_payment_tokens` table,  `token` column)
- Purchase another product, adding the payment method as a new method
- Check the option "Save payment information to my account for future purchases." again
- Confirm that no new card appears on your payment methods page
- Check the database and confirm that just the `token` column was updated with a new payment method ID

### Shortcode checkout
- As a merchant, create a shortcode checkout page
- Repeat the steps above for the shortcode checkout

\* Subscription products will ignore this new behavior to avoid breaking renewals

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

wjrosa committed:
Fix creation of duplicate tokens

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
Fix creation of duplicate tokens

wjrosa committed:
Fix duplicate tokens bug when updateing existing ones

wjrosa committed:
Changelog and readme entries

wjrosa committed:
Removing duplicate method

wjrosa committed:
Reverting unnecessary change

wjrosa committed:
Fix duplicate tokens for the block checkout

wjrosa committed:
Readme and changelog updates

wjrosa committed:
Readme and changelog updates

wjrosa committed:
Not updating tokens if the new order is a subscription

wjrosa committed:
Block existing token update if any subscription is using it

wjrosa committed:
Replacing token update block with new method

diegocurbelo committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
Removing subscription limitation

wjrosa committed:
Comparing fingerprint instead of card details + new fingerprint trait + moving tokens to a new folder

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
Fix CC token class overitte

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Removing unnecessary trait usage

wjrosa committed:
Fix tests

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Adding specific unit tests

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
Fix tests

diegocurbelo reviewed (changes requested):
Thanks for working on this @wjrosa!

It tests well, and the changes look good except for the `$token` type comments, they make the switch block harder to read.
I guess the reason nehind them is the `$token->get_fingerprint()` vs `$token->get_email()` vs `$token->get_cashtag()` used for different payment methods.

Maybe we can extend the trait `WC_Stripe_Fingerprint_Trait` (or the base class 🤷🏼) to handle them all in one place? For example, by having `get_fingerprint()` return the fingerprint/email/cashtag as needed; that way the code in `search_for_duplicate_token()` can be simplified and it makes easier to add a new payment method with a "custom" fingerprint in the future.

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

wjrosa committed:
New trait to simplify duplicate comparison logic

wjrosa committed:
Renaming trait + including it

wjrosa committed:
Specific unit tests

wjrosa committed:
Renaming the search method to get

wjrosa commented:
> Thanks for working on this @wjrosa!
> 
> It tests well, and the changes look good except for the `$token` type comments, they make the switch block harder to read. I guess the reason nehind them is the `$token->get_fingerprint()` vs `$token->get_email()` vs `$token->get_cashtag()` used for different payment methods.
> 
> Maybe we can extend the trait `WC_Stripe_Fingerprint_Trait` (or the base class 🤷🏼) to handle them all in one place? For example, by having `get_fingerprint()` return the fingerprint/email/cashtag as needed; that way the code in `search_for_duplicate_token()` can be simplified and it makes easier to add a new payment method with a "custom" fingerprint in the future.

You're right, Diego. I have created a new trait `WC_Stripe_Token_Comparison_Trait` and simplified the logic. I would prefer a shared base class, but our current token classes do not always share the same base class (`WC_Stripe_Payment_Token_CC` extends `WC_Payment_Token_CC`, while `WC_Payment_Token_CashApp` extends `WC_Payment_Token`, both base classes are WC classes). So this new trait is used by all of them instead.

Edit: I decided to transform the trait into an interface and removed the switch block (each class now implements the method with its own logic).

wjrosa committed:
Transforming trait into interface

wjrosa committed:
Fix tests

diegocurbelo committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens

diegocurbelo reviewed (approved):
Thanks for the changes @wjrosa 

LGTM 🚢

wjrosa committed:
Merge branch 'develop' into fix/duplicate-credit-card-tokens
